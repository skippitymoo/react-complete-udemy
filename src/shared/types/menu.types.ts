export type Meal = {
  /**
   * Unique identifier for the meal item
   */
  id: string;
  /**
   * Name of the meal item
   */
  name: string;
  /**
   * Description for the meal item
   */
  description: string;
  /**
   * Price of the meal item
   */
  price: number;
};

export type MealsAvailable = {
  /**
   * List of meals available
   */
  availableMeals: Meal[];
};

export type CartItem = {
  /**
   * Number of items of this meal
   */
  amount: number;
  /**
   * The meal for this cart item
   */
  meal: Meal;
};

export type CartItemActions = {
  /**
   * Modify the amount for the cart item
   */
  onAmountChange: (mealId: string, changeBy: number) => void;
};

export type Cart = {
  /**
   * Cart items currently in the cart
   */
  initialCartItems: CartItem[];
};

export type CartActions = {
  /**
   * The new total number of items in the cart
   */
  onCartChange: (totalCartItems: number) => void;
  /**
   * Order is placed
   */
  onOrder: () => void;
};

export type CartProps = Cart & CartActions;
