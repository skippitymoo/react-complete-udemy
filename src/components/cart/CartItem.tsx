import { JSX } from 'react';
import { Chip } from '../ui/Chip';
import { Button } from '../ui/Button';
import { CartItem as CartItemType } from '../menu.types';
import { formatCurrency } from '../../utils/utilities';

export const CartItem = ({ amount, meal }: CartItemType): JSX.Element => {
  return (
    <div className='cart-item'>
      <div className='cart-item__details'>
        <div className='cart-item__title'>{meal.name}</div>
        <span className='cart-item__price'>{formatCurrency(meal.price)}</span>
        <Chip className='cart-item__amount'>{`x ${amount}`}</Chip>
      </div>
      <div className='cart-item__actions'>
        <Button size='small' variant='light'>
          &minus;
        </Button>
        <Button size='small' variant='light'>
          +
        </Button>
      </div>
    </div>
  );
};
