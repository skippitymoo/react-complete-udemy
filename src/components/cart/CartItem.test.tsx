import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { CartItem } from './CartItem';
import { Meal } from '../menu.types';

expect.extend(toHaveNoViolations);

describe('CartItem', () => {
  it('renders all the props', async () => {
    const meal: Meal = {
      id: 'dummy',
      name: 'Schnitzel',
      description: 'dummy',
      price: 16.5,
    };
    const { container } = render(<CartItem meal={meal} amount={4} />);

    expect(await axe(container)).toHaveNoViolations();
    expect(container.firstChild).toMatchSnapshot();
  });
});
