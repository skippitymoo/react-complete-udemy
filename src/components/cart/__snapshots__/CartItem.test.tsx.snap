// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem renders all the props 1`] = `
<div
  class="cart-item"
>
  <div
    class="cart-item__details"
  >
    <h3
      class="cart-item__title"
    >
      Schnitzel
    </h3>
    <span
      class="cart-item__price"
    >
      £16.50
    </span>
    <span
      class="chip cart-item__amount"
    >
      x 4
    </span>
  </div>
  <div
    class="cart-item__actions"
  >
    <button
      class="btn btn--light btn--small"
    >
      −
    </button>
    <button
      class="btn btn--light btn--small"
    >
      +
    </button>
  </div>
</div>
`;
